### PR TITLE
mlos_bench config json schema validation - environments

### PR DESCRIPTION
**Pull request by bpkroth**
Extends #340, #346, #349, #352 to Environment configs.
See Also: #331 

- [x] main schema as a set of subschemas
- [x] connect schema validation in the mlos_bench code
- [x] test-cases

### CODE REVIEW

File: mlos_bench/mlos_bench/config/environments/apps/redis/redis.jsonc [hunk 1]
```diff
@@ -1,87 +1,103 @@
-[
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/config/environments/apps/redis/redis.jsonc:1]: Reworks environment list configs into CompositeEnv configs as discussed in #340

File: mlos_bench/mlos_bench/config/schemas/environments/local/local-fileshare-env-subschema.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                            "description": "List of file upload config objects.",
+                            "type": "array",
+                            "items": {
+                                "$ref": "../common-environment-subschemas.json#/$defs/file_upload_config"
+                            }
+                        }
+                    }
+                }
+            ],
+            "unevaluatedProperties": false
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/config/schemas/environments/local/local-fileshare-env-subschema.json:51]: - [ ] Make at least one of `upload` or `download` required.

File: mlos_bench/mlos_bench/config/schemas/environments/local/local-env-subschema.json [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                    "$ref": "../common-environment-subschemas.json#/$defs/leaf_environment_config"
+                },
+                {
+                    "$ref": "../common-environment-subschemas.json#/$defs/setup_run_teardown_configs"
+                },
+                {
+                    "$ref": "#/$defs/local_env_config"
+                }
+            ],
+            "unevaluatedProperties": false
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/config/schemas/environments/local/local-env-subschema.json:76]: Make one of the config properties required.

File: mlos_bench/mlos_bench/tests/config/schemas/optimizers/test_optimizer_schemas.py
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/config/schemas/optimizers/test_optimizer_schemas.py:1]: Expand the set of files we check for this - useful when schemas are split into subschemas to make sure we didn't do something for one and not the other.

File: mlos_bench/mlos_bench/tests/config/schemas/README.md
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/config/schemas/README.md:1]: Documents some of the layout and logic of the test schemes used here.

File: mlos_bench/mlos_bench/config/environments/apps/redis/redis.jsonc (hunk 1 above)
Comment (motus) [reply mlos_bench/mlos_bench/config/environments/apps/redis/redis.jsonc:1]: Just to illustrate, here's a hierarchy that we have now:
```
CompositeEnv :: 'Config for: Azure VM Ubuntu Redis'
    VMEnv :: 'Deploy Ubuntu VM on Azure'
    CompositeEnv :: 'Setup Ubuntu VM'
        OSEnv :: 'Boot Ubuntu VM on Azure',
        LocalFileShareEnv :: 'Generate boot parameters for a Ubuntu VM',
        RemoteEnv :: 'Apply boot parameters to a remote Ubuntu VM'
    CompositeEnv :: 'Setup Ubuntu VM'
        LocalFileShareEnv :: 'Generate Linux kernel parameters for Ubuntu'
        RemoteEnv :: 'Apply Linux kernel parameters to a remote Ubuntu VM'
    CompositeEnv :: 'Redis Composite Env'
        LocalFileShareEnv :: 'Generate Redis configuration'
        RemoteEnv :: 'Benchmark Redis on Linux'
        LocalFileShareEnv :: 'Download and post-process Redis benchmark data'
```
Before that, we used to have a single root `CompositeEnv` with a flat list of child environments underneath.

P.S. I am not saying that the new schema is worse or better, just different

Comment (motus) [reply mlos_bench/mlos_bench/config/environments/apps/redis/redis.jsonc:1]: P.S. I'll probably add a function to print that tree to the log

**Review by motus — APPROVED** [review]
Left a few comments, but overall looks great!

File: mlos_bench/mlos_bench/tests/config/schemas/environments/test_environment_schemas.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+# - for each config, load and validate against expected schema
+
+TEST_CASES = get_schema_test_cases(path.join(path.dirname(__file__), "test-cases"))
+
+
+# Dynamically enumerate some of the cases we want to make sure we cover.
+
+expected_environment_class_names = [subclass.__module__ + "." + subclass.__name__
+                                    for subclass
+                                    in get_all_concrete_subclasses(Environment)]  # type: ignore[type-abstract]
```
[the remark this answers is not in the document]
Comment (bpkroth) [reply mlos_bench/mlos_bench/tests/config/schemas/environments/test_environment_schemas.py:37]: Surprised at that, but ok *shrug*

**Pull request merged**